contents/linux/shell.md

```diff
@@ -0,0 +1,180 @@
+read -p "Please enter a directory name: " dir
+if [ "$dir" == "" -o ! -d "$dir" ]; then
+	echo "The $dir is NOT exist in your system"
+	exit 1
+fi
+
+# output permissions for each of the file under the directory.
+filelist=$(ls $dir)
+for filename in $filelist
+do 
+	perm=""
+	test -r "$dir/$filename" && perm="readable"
+	test -w "$dir/$filename" && perm="$perm writable"
+	test -x "$dir/$filename" && perm="$perm executable"
+	echo "The file $dir/$filename 's permission is $perm" 
+done
+```
+**Functions** in bash script just like functions in normal programming languages, more semantics of bash script can be found [here](https://www.gnu.org/software/bash/manual/html_node/). 
+
+bash script allows us to automate the frequently performed operations, it is easy to use and can be executed on any Unix-like operating systems without any modifications.
+
+### Benefit: Data Stream Redirection
+
+In linux, everything is abstracted as file, including **input**, **output**, and **error** streams. **Redirection** is a feature in Linux such that when executing a command/user program, you can change the standard input/output devices by changing its [file descriptor](https://en.wikipedia.org/wiki/File_descriptor). Each command and user program in linux is an application with input, output and errors. We can redirect the output of one application to the input of another, as such, we combined the two applications together as if we build a pipe between them.
+
+![file-descriptor](images/file-descriptor-illustration.jpg)
+
+For example, suppose you just finished your coding assignment, the resulting executable `calculate` takes a user input, do the calculation, and gives the output. To run it, you simply types:
+
+```bash
+./calculate
+```
+
+But you are not satisfied with manual testing, instead, you want to test with larger data set file `DataBundle`. And verify the output with another executable `verify`, which takes in the result as input and verify its correctness and output **PASS** or **FAIL**. Here's how we can do it.
+
+```bash
+./calculate < DataBundle > result
+./verify < result
+```
+Notice in this case, `calculate` uses the data in `DataBundle` as input, and output the results to the file `result`. Then, `result` is taken as input to `verify`, and the final verification result is printed on screen.
+
+Furthermore, if the code we write is buggy, it throws exception and error messages are directed to error output, then our result file will be empty! To handle it properly, we can also redirect error output to a file `errors`, and verify it is empty(such that no error occurs during execution) before we execute `verify`.
+
+```bash
+#!/bin/bash
+./calculate < DataBundle > result 2> errors
+if [ -s errors]
+then
+     ./verify < result
+else
+     echo "error occured during execution."
+fi
+```
+Another useful command is `|`, which chained linux commands together, as such the output of previous commands is passed as input to the next command.
```

Comment:
one nice thing to mention in here could be `tee`, and since you used the different streams here, might wanna put some link to it with more explanation?